File: 백준/Gold/13023. ABCDE/ABCDE.py
from sys import stdin, setrecursionlimit

setrecursionlimit(10 ** 5)
input = stdin.readline

n, m = map(int, input().split())  # n: 사람의 수, m: 관계의 수
vertex = [[] for _ in range(n)]

for _ in range(m):
    a, b = map(int, input().split())  # a <-> b: 친구 관계
    vertex[a].append(b)
    vertex[b].append(a)


def dfs(x):
    global cnt, flag
    if cnt >= 5:
        flag = True
        return

        # 한 경로에 있는 요소 개수를 파악하기 위해선 복원 처리가 필요
    for nxt in vertex[x]:
        if visited[nxt]:
            cnt += 1
            visited[nxt] = False  # 방문 처리
            dfs(nxt)
            visited[nxt] = True  # 복원 처리
            cnt -= 1


visited = [True] * n
flag = False
for i in range(n):
    # 방문이 가능하다면 -> dfs 진행
    if visited[i]:
        visited[i] = False  # 방문 처리
        cnt = 1
        dfs(i)
        visited[i] = True  # 복원 처리
        if flag:
            break

if flag:
    print(1)
else:
    print(0)

File: Java/백준/Gold/13023. ABCDE/ABCDE.java
import java.io.*;
import java.util.*;

public class Main {

  static boolean flag;
  static int N, M, a, b;
  static boolean[] visited;
  static ArrayList<ArrayList<Integer>> vertex;

  public static void main(String[] args) throws IOException {
    init();
    solve();
  }

  private static void init() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;

    st = new StringTokenizer(br.readLine());
    N = Integer.parseInt(st.nextToken());
    M = Integer.parseInt(st.nextToken());

    vertex = new ArrayList<>();
    for (int i = 0; i < N; i++) {
      vertex.add(new ArrayList<>());
    }

    for (int i = 0; i < M; i++) {
      st = new StringTokenizer(br.readLine());
      a = Integer.parseInt(st.nextToken());
      b = Integer.parseInt(st.nextToken());
      // a <-> b: 양방향 관계
      vertex.get(a).add(b);
      vertex.get(b).add(a);
    }

    br.close();
  }

  private static void solve() {

    flag = false;
    for (int i = 0; i < N; i++) {
      visited = new boolean[N];
      visited[i] = true; // 방문 처리
      dfs(i, 1);
      if (flag) {
        break;
      }
    }

    System.out.println(flag ? 1 : 0);
  }

  private static void dfs(int v, int cnt) {
    if (cnt == 5) {
      flag = true;
      return;
    }
    for (Integer nxt : vertex.get(v)) {
      // 이미 방문한 적이 있다면 탐색하지 않는다.
      if (visited[nxt]) {
        continue;
      }
      visited[nxt] = true; // 방문 처리
      dfs(nxt, cnt + 1);
      visited[nxt] = false; // 복원 처리
    }
  }
}
